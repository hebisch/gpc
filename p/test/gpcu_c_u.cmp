#!/bin/sh
BASEDIR="`echo $2 | sed -e 's,[^/]*$,,'`"
COMPILE_NO_LINK="`echo $1 | sed -e 's/ -o *[^ ]*//'` -c"
$COMPILE_NO_LINK "$BASEDIR"/gpcu_dum.pas #-o gpcu_dum.o
rm -f "$A_OUT"
$COMPILE_NO_LINK $2
$PC_WITHOUT_GP -o "$A_OUT" gpcu_c_c.o gpcu_c_u.o  # with GP
if [ -r "$A_OUT" ]; then
  ./"$A_OUT" > testmake.tmp
  if [ x"$FIXCR" = x ]; then
    diff testmake.tmp "$BASEDIR"/gpcu_c_u.out
    stat=$?
  else
    "$FIXCR" < testmake.tmp > diff_cr1.tmp
    "$FIXCR" < "$BASEDIR"/gpcu_c_u.out > diff_cr2.tmp
    diff diff_cr1.tmp diff_cr2.tmp
    stat=$?
    rm -f diff_cr1.tmp diff_cr2.tmp
  fi
  if [ $stat -eq 0 ]; then
    echo "OK"
  else
    echo "failed"
  fi
else
  echo "failed"
fi
