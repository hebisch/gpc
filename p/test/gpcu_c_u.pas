{ COMPILE-CMD: gpcu_c_u.cmp }

{ We lie to the test script that this is a PROGRAM. ;-}

{ GPC demo unit to show how to include Pascal code in C programs.

  Copyright (C) 2000-2006 Free Software Foundation, Inc.

  Author: Frank Heckenbach <frank@pascal.gnu.de>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License as
  published by the Free Software Foundation, version 2.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; see the file COPYING. If not, write to
  the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
  Boston, MA 02111-1307, USA.

  As a special exception, if you incorporate even large parts of the
  code of this demo program into another program with substantially
  different functionality, this does not cause the other program to
  be covered by the GNU General Public License. This exception does
  not however invalidate any other reasons why it might be covered
  by the GNU General Public License. }

unit GPCU_C_Unit;

interface

{ External declarations we use from the C code }

var
  CVariable : Integer; external name 'c_variable';

procedure CRoutine; external name 'c_routine';

{ Pascal code }

var
  PascalUnitVariable : Integer = 42; attribute (name = 'pascal_unit_variable');

procedure PascalUnitRoutine; attribute (name = 'pascal_unit_routine');

implementation

procedure PascalUnitRoutine;
begin
  WriteLn ('Routine in Pascal unit called from C code.');
  WriteLn ('PascalUnitVariable is now ', PascalUnitVariable, '.');
  Flush (Output);

  WriteLn ('Decrementing CVariable.');
  Dec (CVariable);

  WriteLn ('Calling CRoutine.');
  Flush (Output);
  CRoutine;

  WriteLn ('Back in Pascal unit''s routine');
  Flush (Output)
end;

to begin do
begin
  WriteLn ('Initializer of Pascal unit called.');
  WriteLn ('PascalUnitVariable is now ', PascalUnitVariable, '.');
  Flush (Output)
end;

to end do
begin
  WriteLn ('Finalizer of Pascal unit called.');
  Flush (Output)
end;

end.
