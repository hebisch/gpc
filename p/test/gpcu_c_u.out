We start in the C `main'.
Initializing the Pascal RTS.
Calling the Pascal initializers.
Initializer of Pascal unit called.
PascalUnitVariable is now 42.
Back in C `main'.
Incrementing pascal_unit_variable.
c_variable is 42.
Calling pascal_unit_routine.
Routine in Pascal unit called from C code.
PascalUnitVariable is now 43.
Decrementing CVariable.
Calling CRoutine.
C routine called from Pascal code.
c_variable is now 41.
Back in Pascal unit's routine
Back in C `main'.
Calling the Pascal finalizer.
Finalizer of Pascal unit called.
Done.
